pkg/uploader/provider/kopia.go

```diff
@@ -0,0 +1,204 @@
+/*
+Copyright The Velero Contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package provider
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/kopia/kopia/snapshot/snapshotfs"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	"github.com/vmware-tanzu/velero/pkg/uploader"
+	"github.com/vmware-tanzu/velero/pkg/uploader/kopia"
+
+	"github.com/vmware-tanzu/velero/internal/credentials"
+	velerov1api "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+	repokeys "github.com/vmware-tanzu/velero/pkg/repository/keys"
+	"github.com/vmware-tanzu/velero/pkg/repository/udmrepo"
+	"github.com/vmware-tanzu/velero/pkg/repository/udmrepo/service"
+)
+
+//BackupFunc mainly used to make testing more convenient
+var BackupFunc = kopia.Backup
+var RestoreFunc = kopia.Restore
+
+//kopiaProvider recorded info related with kopiaProvider
+type kopiaProvider struct {
+	bkRepo        udmrepo.BackupRepo
+	credGetter    *credentials.CredentialGetter
+	uploader      *snapshotfs.Uploader
+	restoreCancel chan struct{}
+	log           logrus.FieldLogger
+}
+
+//NewKopiaUploaderProvider initialized with open or create a repository
+func NewKopiaUploaderProvider(
```

Comment:
Isn't the `velerov1.BackupRepo` a parameter of the function? Or how does the uploader know where to put the files?

And seems only the BSL's UID is used, how does the uploader know the access information(e.g. URL) of the storage?

Reply:
Yes, I am considering this problem, we need to use BSL+VolumeNamespace to identify a repository. On the other hand, actually, if we have the `velerov1.BackupRepo`, it is enough to just use `velerov1.BackupRepo`'s UUID to identify a repository, which is quite simple. However, in the current code, PVB/PVR doesn't get the `velerov1.BackupRepo`.

Let me consider this for some more time and discuss with you later. In the current PR, let's make it as is.

Reply:
Let me add some more explanation here:

- During the Connect operation of unified repo, a config file containing all the connecting information is stored in a local file
- During the backup/restore, as long as the uploader could provide the name of the config file, it could connect to the repo and then write/read data to/from the repo
- The Connection operation is done during repo initialization, that is, by the `velerov1.BackupRepo` reconcile
- We need a unique file name for the local config so that different repo has a unique config
- At present, the config file is named by the BSL's UUID

As mentioned above BSL's UUID is not enough, `velerov1.BackupRepo`'s UUID is more suitable, however, it is not currently available in PVB/PVR

Reply:
yeah, I've changed into `velerov1.BackupRepo`'s UUID